### DISCUSSION TITLE
How UnoCSS generate dynamic rule

### DISCUSSION THREAD

**Discussion by ghost** [unocss/unocss#3011] [Q&A]
Hello, thank you for the fantastic works. Before I decide to use UnoCSS on my project, I've been using hours to read through the doc and the story behind it. But I still cannot understand how it work behind the scene...

It said:

> No parsing, no AST, no scanning. It's 5x faster than Windi CSS or Tailwind JIT.

This is the most comfusing part for me, without scanning the source code / HTML, how do UnoCSS generate the CSS that is present in the HTML? At first, this seems logic to me if it's static. For example:
`
rules: [
    ['m-1', { margin: '0.25rem' }],
]
`

UnoCss then generate `.m-1 { margin: 0.25rem; }` based on this config, therefore scanning is not necessary. But how about dynamic rule like `[/^m-(\d+)$/, ([, d]) => ({ margin: `${d / 4}rem` })]`, no matter what it still somewhat need to 'scan' the HTML, else there is no way UnoCSS will know what are the present CSS are in the HTML like `m-100`? The '100' is a dynamic value, without looking / scanning the HTML, how UnoCSS get this value?

Here, do I misunderstood the meaning of 'scanning'? Or UnoCSS actually did scanning?

P.S.: I'm planning to use the CDN with 'Core', then create my own preset. In other word, it's on the runtime and not the build time.

**Comment by Shyam-Chen** [comment]
> UnoCSS will read the content that goes through your build tools pipeline and extract the utilities usages from them.

- `/\.(vue|svelte|[jt]sx|mdx?|astro|elm|php|phtml|html)($|\?)/,`
- `@unocss-include`

**Comment by ghost** [comment]
@Shyam-Chen  That's the reason I wrote the postscript...

> goes through your build tools pipeline

This seems logic if I'm using CLI which run at build time. But I'm using the runtime 'CDN'. Without scanning the HTML at runtime ( CDN ), how do UnoCSS know / get the present CSS in order to generate them?

**Comment by zyyv** [comment]
@karchung0930 Hi, as https://unocss.dev/guide/extracting#extracting-from-build-tools-pipeline said, `Uno` will extract the utility from the source code of the qualified file, analyze and generate css.

**Comment by ghost** [comment]
@zyyv 

> UnoCSS works by searching for the utilities usages from your codebase

For runtime, I believe the 'searching' is actually scanning...No? Haha

The content source 'Pipeline' and 'Filesystem' from your reference are relevant only for build time. Here I have mentioned two times that I'm using the runtime CDN.

In conclude, UnoCSS still did scanning for runtime like using with CDN. Your doc should specify that 'No scanning for build time' or just remove 'remove scanning'.

**Reply by zyyv** [reply]
https://unocss.dev/integrations/runtime We have runtime package, but I don't know what it has to do with you building your own preset, you just need to pay attention to your rules or variants ...

**Reply by ghost** [reply]
To be honest, I still don't get it...Sorry, haha. Anyway, I'll use it without fully understand for now!
